File: leetcode/209.min-sub-array-len/ts/minsubarraylen.ts
const minSubArrayLen = (target: number, nums: number[]): number => {
    return minSubArrayLenOptimized(target, nums)
};

// N, 1
const minSubArrayLenOptimized = (target: number, nums: number[]): number => {
    let minimumLength = nums.length + 1
    let l = 0
    let sum = 0

    for (let r = 0; r < nums.length; r++) {
        sum += nums[r]

        while (sum >= target) {
            const len = r - l + 1
            if (len < minimumLength) minimumLength = len

            sum -= nums[l]
            l++
        }
    }

    return minimumLength === nums.length + 1 ? 0 : minimumLength
};

export default minSubArrayLen

File: leetcode/209.min-sub-array-len/cpp/minsubarraylen.cpp
#include <vector>

class Solution
{
public:
    int minSubArrayLen(int target, std::vector<int> &nums)
    {
        return minSubArrayLenOptimized(target, nums);
    }

private:
    // N, 1
    int minSubArrayLenOptimized(int target, std::vector<int> &nums)
    {
        int left = 0;
        int sum = 0;
        int res = nums.size() + 1;

        for (int right = 0; right < nums.size(); right++)
        {
            sum += nums[right];

            while (sum >= target)
            {
                const int curr = right - left + 1;
                if (curr < res)
                    res = curr;

                sum -= nums[left];
                left++;
            }
        }

        return res == nums.size() + 1 ? 0 : res;
    }
};
